### ISSUE TITLE
Huge time taken by pysecdec for prefactor expansion

### ISSUE THREAD

**Issue by gkdgoutam** [gudrunhe/secdec#23]
Hello,

I am experiencing  a huge time taken in pysecdec just for the expansion for the prefactor:

`prefactor='-cos(2*pi*eps)*pi*cot(pi*eps)*exp(-2*EulerGamma*eps) * (2**(1-4*eps))/(eps**2)*csc(pi*eps) * gamma(1-4*eps)*(gamma(1-eps))**2/(gamma(1-2*eps))**2 *1/(1-eps)/(2*eps-1)'`

`expanding the prefactor -1.0*2**(1 - 4*eps)*pi*exp(-2*EulerGamma*eps)*cos(2*pi*eps)*cot(pi*eps)*csc(pi*eps)*gamma(1 - 4*eps)*gamma(1 - eps)**2/(eps**2*(1 - eps)*(2*eps - 1)*gamma(1 - 2*eps)**2) (regulators: [alp, eps] , orders: [0 0] )`
and it is stuck for hours! Without the prefactor everything is as usual and very fast!

The expression does not take time to expand in a python shell with sympy(it's instantaneous up to terms eps^1). 
However, inside the pysecdec framework, it seems to take a colossal time: something between 5mins - 120mins!!

**Is it due to the fact that I have two regulators (although the prefactor depends only on one)?**
Is there any insight on this?

NB: The expression can be simplified before putting it inside pysecdec, which is unwanted (but also does not help much).

pysecdec : 1.5.3
sympy: 1.6 

I can upgrade sympy to 1.10.1 but that would lead to an error that some functions are deprecated! But that also does not improve the time. Here is a snap of that with sympy 1.10.1:

``sympify() performed the string fallback resulting in the following string:

' + (0)'

`
See https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-sympify-string-fallback
for details.

This has been deprecated since SymPy version 1.6. It
will be removed in a future version of SymPy.

  return sp.sympify(a)
/home/users/goutam/.local/lib/python3.7/site-packages/pySecDec/misc.py:439: SymPyDeprecationWarning:

The string fallback in sympify() is deprecated.

To explicitly convert the string form of an object, use
sympify(str(obj)). To add define sympify behavior on custom
objects, use sympy.core.sympify.converter or define obj._sympy_
(see the sympify() docstring).`

Many thanks,
GD

**Comment by spj101** [comment]
To your first question, I suspect this is a [known issue with sympy](https://github.com/sympy/sympy/issues/2635#issuecomment-284434997). The (incredibly stupid) solution might be to manually use the identity `gamma(1-x) = -x*gamma(-x)`. 

We specifically tried to workaround this deprecation in a75fa0c3 it is not immediately clear to me why line 440 of misc.py is not catching this warning.

**Comment by spj101** [comment]
Without hacking `make_package` I don't see a way to tell pySecDec not to expand the prefactor in both variables and we would need to be careful not to mess up that pySecDec expects a polynomial in both variables (even if we only have an `alp**0` term). The way we do the series expansion with sympy is term by term (calling `series(...,n=None)` and then `next()`). This seems to be especially slow on your example.

One nasty hack could be to pre-expand:
```
expr = -cos(2*pi*eps)*pi*cot(pi*eps)*exp(-2*EulerGamma*eps) * (2**(1-4*eps))/(eps**2)*csc(pi*eps) * gamma(1-4*eps)*(gamma(1-eps))**2/(gamma(1-2*eps))**2 *1/(1-eps)/(2*eps-1)
expr2=sp.series(expr,n=0)
perfactor = expr2
```
Of course, this has the danger that you don't expand high enough and end up truncating your result by accident. If I don't see a smarter way to improve how we call sympy this could be something we do internally...

The bottom line is that series expansions are very slow in sympy (and getting slower with recent versions), the way we need to call sympy seems to be even slower than other ways of asking it to do the series expansion.

**Comment by gkdgoutam** [comment]
Agree. This would be the last resort (maybe expanding in Mathematica to higher orders (I don't like this though))! Is there any possibility to drop sympy completely and use eg. FORM to even do this prefactor expansion? I can think of applicability will be one issue, not all functions are available out of the box in FORM, but it will be no doubt much faster!

**Comment by spj101** [comment]
FORM does not have series expansions built-in by default. We would have to implement them for each function by hand, your example already contains several special functions (`cot`,`csc`) which we do not normally encounter.

It would be so convenient if sympy had reasonable performance but I agree this is not really acceptable.

**Comment by gkdgoutam** [comment]
Trigonometric functions can be related to `gamma`s. But I agree one has to write some specific FORM routines (but probably not much)!

**Comment by gkdgoutam** [comment]
> We specifically tried to workaround this deprecation in [a75fa0c](https://github.com/gudrunhe/secdec/commit/a75fa0c38517c7669831abe83859112ba9fa2fe7) it is not immediately clear to me why line 440 of misc.py is not catching this warning.


FYI, this appears only in sympy1.10.1.    Version 1.9 is fine.

**Comment by spj101** [comment]
We have migrated prefactor expansion from sympy to ginac following [sympy #24266](https://github.com/sympy/sympy/issues/24266) and other, more general, issues you encountered here.

The new ginac prefactor expansion is available in our `proposed` branch. There could be some issues/bugs related to the incompatible syntax of the two programs and we are happy to hear about any issues our users encounter.
